internal/v4/api.go

```diff
@@ -1169,6 +1170,61 @@ func (h *ReqHandler) serveAdminPromulgate(id *router.ResolvedURL, w http.Respons
 	return nil
 }
 
+// PUT id/publish
+// See https://github.com/juju/charmstore/blob/v4/docs/API.md#put-idpublish
+func (h *ReqHandler) servePublish(id *charm.URL, w http.ResponseWriter, req *http.Request) error {
+	// Perform basic validation of the request.
+	if req.Method != "PUT" {
+		return errgo.WithCausef(nil, params.ErrMethodNotAllowed, "%s not allowed", req.Method)
+	}
+	if id.Channel == charm.DevelopmentChannel {
```

Comment:
if id.Channel != "" {

might be a little more future-proof. (the error message might be "can only set publish on published URL" or something like that)